src/compiletest.md

```diff
@@ -0,0 +1,183 @@
+code.
+
+#### Using a header command
+Here is an example, specifying the `must-compile-successfully` header command, which takes no arguments, followed by the
+`failure-status` header command, which takes a single argument (which, in this case is a value of 1).  `failure-status` is
+instructing `compiletest` to expect a failure status of 1 (rather than the current Rust default of 101 at the time of this
+writing).  The header command and the argument list (if present) are typically separated by a colon:
+```
+// Copyright 2018 The Rust Project Developers. See the COPYRIGHT
+// file at the top-level directory of this distribution and at
+// http://rust-lang.org/COPYRIGHT.
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+// must-compile-successfully
+// failure-status: 1
+
+#![feature(termination_trait)]
+
+use std::io::{Error, ErrorKind};
+
+fn main() -> Result<(), Box<Error>> {
+    Err(Box::new(Error::new(ErrorKind::Other, "returned Box<Error> from main()")))
+}
+```
+
+#### Adding a new header command property
+One would add a new header command if there is a need to define some test property or behavior on an individual, test-by-test
+basis.  A header command property serves as the header command's backing store (holds the command's current value) at
+runtime.
+
+To add a new header command property:
+    1. Look for the `pub struct TestProps` declaration in `src/tools/compiletest/src/header.rs` and add
+the new public property to the end of the declaration.
+    2. Look for the `impl TestProps` implementation block immediately following the struct declaration and initialize the new
+property to its default value.
+
+#### Adding a new header command parser
+When `compiletest` encounters a test file, it parses the file a line at a time by calling every parsers defined in the
+`Config` struct's implementation block, also in `src/tools/compiletest/src/header.rs` (note the `Config` struct's declaration
+block is found in `src/tools/compiletest/src/common.rs`).  `TestProps`'s `load_from()` method will try passing the current
+line of text to each parser, which, in turn typically checks to see if the line begins with a particular commented (`//`)
+header command such as `// must-compile-successfully` or `// failure-status`.  Whitespace after the comment marker is
+optional.
+
+Parsers will override a given header command property's default value merely by being specified in the test file as a header
+command or by having a parameter value specified in the test file, depending on the header command.
+
+Parsers defined in `impl Config` are typically named `parse_<header_command>` (note kebab-case `<header-command>` transformed
```

Comment:
lol! I have never heard it called kebab-case before 😂

Reply:
:)